### PR TITLE
Feature auth

### PR DESCRIPTION
**Pull request by prateek041**
This pull request will add auth and user management

### CODE REVIEW

File: server/src/schema/mutation/User.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          email: args.email,
+          password: HashedPass,
+          profile: NewUserProfile._id
+        })
+
+        await NewUser.save();
+
+        // Maybe I should change this response
+        return "User created succesfully"
+      }
```
Comment (barelyhuman) [inline server/src/schema/mutation/User.js]: add cleanup for if either of the above functions fail, or use mongoose transactions when inserting into multiple collections

File: server/src/schema/query/query.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+const ProfilesCollection = require('../../models/Profiles')
+const UsersCollection = require("../../models/Users")
+const {GraphQLString} = require("graphql")
+
+const {GraphQLList} =require("graphql")
```
Comment (barelyhuman) [inline server/src/schema/query/query.js]: could be a single `require`

File: server/src/schema/types/types.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const {
+  GraphQLObjectType,
+  GraphQLID,
+  GraphQLString,
+} = require('graphql');
+
+const UserType = new GraphQLObjectType({
```
Comment (barelyhuman) [inline server/src/schema/types/types.js]: Keep the types and resolvers next to each other, each entity should export it's own resolver and types. 

They can be in the different files in the same folder but exported and considered as one entity.

File: server/src/utils/hashingUtil.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const bcrypt = require('bcrypt');
+
+const saltCount = process.env.SaltRound;
+
+const HashPass = async(password)=>{
```
Comment (barelyhuman) [inline server/src/utils/hashingUtil.js]: `camelCase` function names. `PascalCase` is to be used for constructor functions.

File: server/src/schema/schema.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+// collection of queries
+const {GetAllUsersProfile, GetUserProfile} = require("../schema/query/query")
+
+const {
+  GraphQLObjectType,
+  GraphQLSchema,
+} = require('graphql');
+
+// RootQuery type
+const RootQuery = new GraphQLObjectType({
```
Comment (barelyhuman) [inline server/src/schema/schema.js]: could be named `queries` ? 
maintain consistent naming

File: server/src/schema/query/query.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+const {UserType} = require("../types/types")
+
+// Get all users
+const GetAllUsersProfile = {
```
Comment (barelyhuman) [inline server/src/schema/query/query.js]: `camelCase`

File: server/src/index.js [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
+// Connection to the database
+connectDB();
+
 // Middleware to implement GraphQL
 app.use("/gql", graphqlHTTP({
-  schema: typeDefs,
-  rootValue: resolvers,
+  schema: schema,
   graphiql: true
 }))
```
Comment (barelyhuman) [inline server/src/index.js:24]: mongoose and mongo do not have immediate connections. 
In certain cases the initial connection might take longer, do not start the `.listen` on the server unless connected. 


provide `connectDB()` with a callback that waits for `mongoose.connection.on("connected",callback)` or add a `await mongoose.connect()` in your db file to actually let the connection finish and await that in this file as well.

**Review by barelyhuman — COMMENTED** [review]
Expected you to use GraphQL DSL instead of writing in Javascript. 
Anyway, since it's done with Javascript and not TS, lookup the docs on how to add Decorators to the schema. 

Other issues 
- no CORS blockage 
- no addition of security headers, add in `helmet` for basic security 
- add in a rate-limiter to the resolvers , look for graphql specific rate limiters 

Everything else LGTM for now

File: server/src/schema/query/query.js (hunk 2 above)
Comment (prateek041) [reply server/src/schema/query/query.js]: resolved this.

File: server/src/schema/types/types.js (hunk 3 above)
Comment (prateek041) [reply server/src/schema/types/types.js]: changed the project structure to better define the schema.

File: server/src/schema/schema.js (hunk 5 above)
Comment (prateek041) [reply server/src/schema/schema.js]: fixed this.

**Comment by prateek041** [comment]
Changed schema definition from GraphQL JS to GraphQL DSL. Other tasks still remaining.

Currently working on finishing User management.

**Pull request merged**